Window to worker communication series (#4340)
GoogleChrome/web.dev: src/site/content/en/reliable/imperative-caching-guide/index.md

tags:
  - service-worker
  - performance
  - offline
---

Some websites might need to communicate to the service worker without the need of being
informed about the result. Here are some examples:

- A page sends the service worker a list of URLs [to
  prefetch](https://web.dev/instant-navigation-experiences/), so that, when the user clicks on a
  link the document or page subresources are already available in the cache, making subsequent
  navigation much faster. 
- The page asks the service worker to retrieve and cache a set of top articles, to have them
  available for offline purposes.

Delegating these types of non-critical tasks to the service worker has the benefit of freeing up the
main thread for better handling more pressing tasks such as responding to user interactions.

<figure class="w-figure">
  <img src="imperative-caching-diagram-1.png"
       width="565"
       alt="Diagram of a page requesting resources to cache to a service worker.">
</figure>

In this guide we'll explore how to implement a **one-way** communication technique from the page to
the service worker by using standard browser APIs and the Workbox library. We'll call these types of

> use cases **imperative caching**.
> 
> ## Production case {: #production-case }

Reviewer: ```suggestion
use cases **imperative caching**.

{% Aside %}
  Check out [Workers overview](/workers-overview/) for a high-level explanation
  of when to use web workers versus service workers and the rest of the
  [Window to workers communication](/reliable/#window-to-worker-communication) series
  for guides on other common use cases.
{% endAside %}

## Production case {: #production-case }
```

- A page sends the service worker a list of URLs [to
  prefetch](https://web.dev/instant-navigation-experiences/), so that, when the user clicks on a
  link the document or page subresources are already available in the cache, making subsequent
  navigation much faster. 
- The page asks the service worker to retrieve and cache a set of top articles, to have them
  available for offline purposes.

Delegating these types of non-critical tasks to the service worker has the benefit of freeing up the
main thread for better handling more pressing tasks such as responding to user interactions.

<figure class="w-figure">
  <img src="imperative-caching-diagram-1.png"
       width="565"
       alt="Diagram of a page requesting resources to cache to a service worker.">
</figure>

In this guide we'll explore how to implement a **one-way** communication technique from the page to
the service worker by using standard browser APIs and the Workbox library. We'll call these types of
use cases **imperative caching**.

{% Aside %}
  Check out [Workers overview](/workers-overview/) for a high-level explanation
  of when to use web workers versus service workers and the rest of the

>   [Window to workers communication](/reliable/#window-to-worker-communication) series

Reviewer: ```suggestion
  [Communicate with workers](/reliable/#communicate-with-workers) series
```

---
layout: post 
title: "Imperative caching" 
authors:
  - demianrenzulli
  - andrewguan 
date: 2020-12-04 
description: | 
  How to communicate window and service worker to perform tasks related to performance, caching and offline. 
tags:
  - blog
  - service-worker
  - offline
  - caching
---

> Some web apps/sites might need to communicate to the service worker without the need of being

Author: Done

informed about the result. Here are some examples:

- A page sends the service worker a list of URLs [to
  prefetch](https://web.dev/instant-navigation-experiences/), so that, when the user clicks on a
  link the document or page subresources are already available in the cache, making subsequent
  navigation much faster. 
- The page asks the service worker to retrieve and cache a set of top articles, to have them
  available for offline purposes.

Delegating these types of non-critical tasks to the service worker has the benefit of freeing up the
main thread for better handling more pressing tasks such as responding to user interactions.

<figure class="w-figure">
  <img src="imperative-caching-diagram-1.png"
       width="565"
       alt="Diagram of a page requesting resources to cache to a service worker.">
</figure>

In this guide we'll explore how to implement a **one-way** communication technique from the page to
the service worker by using standard browser APIs and the Workbox library. We'll call these types of
use cases **imperative caching**.

## Production case

> 1-800 Flowers implemented **imperative caching** (prefetching) with service workers via

Author: Done

> [postMessage()](https://developer.mozilla.org/en-US/docs/Web/API/Worker/postMessage) to prefetch the

Author: Done

top items in category pages to speed up subsequent navigation to product detail pages.

<figure class="w-figure">
  <img src="1-800-flowers-logo.png"
       width="400"
       alt="Logo of 1-800 Flowers.">
</figure>

They use a mixed approach to decide which items to prefetch:

- At page load time they ask the servicer worker to retrieve the JSON data for the top 9 items, and

>   add the resulting response objects to an API cache.

Author: It tried to say a "cache for API calls", but I changed it to just "cache", for simplicity.

> - For the remaining items, they listen to the [mouseover

Author: What is the difference here? They look the same for me.

  event](https://developer.mozilla.org/en-US/docs/Web/API/Element/mouseover_event), so that, when a
  user moves the cursor on top of an item, they can trigger a fetch for the resource on “demand”.

> They use the [cache API](https://developer.mozilla.org/en-US/docs/Web/API/Cache) to store JSON

Author: Done

responses:

<figure class="w-figure">
  <img src="1-800-flowers-prefetching.png"
       width="728"
       alt="Logo of 1-800 Flowers.">
   <figcaption class="w-figcaption">Prefetching JSON product data from product listing pages in 1-800Flowers.com.</figcaption>
</figure>

When the user clicks on an item, the JSON data associated with it can be picked up from the cache,
without the need of going to the network, making the navigation faster.

## Using Workbox

[Workbox](https://developers.google.com/web/tools/workbox) provides an easy way to send messages to

> a service worker, via the [Workbox Window

Author: Done

package](https://developers.google.com/web/tools/workbox/modules/workbox-window), a set of modules

> that are intended to run in the window context. They're a complement to the other workbox packages

Author: Done

that run in the service worker.

To communicate the page with the service worker, first obtain a Workbox object reference to the
registered service worker:

```javascript
const wb = new Workbox('/sw.js');
wb.register();
```

Then you can directly send the message declaratively, without the hassle of the getting the
registration, checking for activation or thinking about the underlying communication API, such as
this:

```javascript
wb.messageSW({"type": "PREFETCH", "payload": {"urls": ["/data1.json", "data2.json"]}}); });
```

> The service worker implements a [“message”

Author: Done

can be used to establish a **one-way** communication mechanism from the page to the service worker.

The page calls
[postMessage()](https://html.spec.whatwg.org/multipage/workers.html#dom-worker-postmessage) on the
service worker interface:

```javascript
navigator.serviceWorker.controller.postMessage({
  type: 'MSG_ID',
  payload: 'some data to perform the task',
});
```
The service worker implements a [“message”
handler](https://developer.mozilla.org/en-US/docs/Web/API/ServiceWorkerGlobalScope/message_event) to
listen to these messages.

```javascript
self.addEventListener('message', (event) => {
  if (event.data && event.data.type === MSG_ID) {
    // do something
  }
});
```

The `{type : ‘MSG_ID'}` attribute is not absolutely required, but it is one way to allow the page to
send different types of instructions to the service worker (i.e. 'to prefetch' vs. 'to clear
storage').  The service worker can branch into different execution paths based on this flag.

In these types of operations there's no need for the service worker to communicate back with the
page. If the operation was successful, the user will be able to get the benefits from it (e.g.

> faster navigation, or offline), but, if not, it won't alter the main user flow.

Author: Done

array of URLs issue a fetch request for each of them:

```javascript
async function fetchAsync(url) {
  // await response of fetch call
  let prefetched = await fetch(url);
  // (optionally) cache resources in the service worker storage
}
```

When the response is obtained you can rely on the caching headers of the resource. In many cases
though, like in product detail pages, resources are not cached (which means, they have a
`Cache-control` header of `no-cache`). In cases like these you can override this behavior, by
storing the fetched resource in the service worker cache. This has the added benefit of allowing the
file to be served in offline scenarios.

### Beyond JSON data

Once the JSON data is fetched from a server endpoint, it often contains other URLs that are also
worth to prefetch, such as an image or other endpoint data that are associated with this first-level
data.  
Let's say that in our example, the JSON data returned is the information of a grocery shopping site,
we'll have something like this:

```json
{
    "productName":"banana",
    "productPic":"https://cdn.example.com/product_images/banana.jpeg",
    "unitPrice":"1.99"

>  }

Author: Done

```

Modify the `fetchAsync()` code to iterate over the list of products and cache the hero image for
each of them:

```javascript
async function fetchAsync(url, postProcess) {
  // await response of fetch call
  let prefetched = await fetch(url);

  //(optionally) cache resource in the service worker cache

  // carry out the post fetch process if supplied
  if (postProcess) {
    await postProcess(prefetched);
  }
}

async function postProcess(prefetched) {
  let productJson = await prefetched.json();
  if (productJson && productJson.product_pic) {
    fetchAsync(productJson.product_pic);
  }
}
```

You can add some exception handling around this code in case situations like server responds a 404
arises. But the beauty of using a service worker to prefetch is that it can fail without much

> consequences to the page and the main thread. You may also have more elaborate logic in the post
> processing of the prefetched content, making it more flexible and decoupled with the data it's

Reviewer: ```suggestion
consequence to the page and the main thread. You may also have more elaborate logic in the
post-processing of the prefetched content, making it more flexible and decoupled with the data it's
```

handling. The sky's the limit.

{% Aside 'caution' %} Prefetching techniques consume extra bytes for resources that are not
immediately needed, so it needs to be applied thoughtfully; only prefetch resources when you are

> confident that users will need them. Consider not prefetching when users are on slow connections.

Reviewer: ```suggestion
confident that users will need them. Avoid prefetching when users are on slow connections.
```
Let's make this a little more forceful

You can detect that with the [Network Information
API](https://developer.mozilla.org/en-US/docs/Web/API/Network_Information_API). {% endAside %}

## Conclusion

In this article, we covered a common use case of **one-way** communication between page and service
worker: **imperative caching**. The examples discussed are only meant for demonstrating one way of
using this pattern and the same approach can be applied to other use cases as well, for example,
caching top articles on demand for offline consumption, bookmarking, and others.

> For more patterns of page and service worker communication, check out:`

Reviewer: ```suggestion
For more patterns of page and service worker communication, check out:
```
